### ISSUE TITLE
Can't draw a line without moving my list of points

### ISSUE THREAD

**Issue by akriegman** [nannou-org/nannou#794]
I have a `Vec<Point2>` in my model and I would like to draw it using `draw.polyline()`. The problem is, `draw.polyline().points(points: I)` accepts an `I: IntoIterator` with `I::Item: Into<Point2>`. As far as I can tell such an `I` would own the underlying data, and so there is no way to get this from an `&Vec` without cloning. Here is a minimal example, notice where I am forced to clone `model.draw_list`:

```
use nannou::prelude::*;

fn main() {
    nannou::app(model)
        .view(view)
        .run();
}

struct Model {
    draw_list: Vec<Point2>
}

fn model(app: &App) -> Model {
    app.new_window()
        .size(720, 720)
        .build()
        .unwrap();

    Model {
        draw_list: vec![pt2(0, 0), pt2(10, 10)]
    }
}

fn view(app: &App, model: &Model, frame: Frame) {
    frame.clear(SKYBLUE);
    let draw = app.draw();
    draw.polyline()
        .weight(2.)
        .points(model.draw_list.clone());
    draw.to_frame(app, &frame).unwrap();
}
```

If we changed the signature of `points` from
```
fn points<I>(self, points: I)
where
    I: IntoIterator,
    I::Item: Into<Point2>,
```
to
```
fn points<'a, I>(self, points: I)
where
    I: IntoIterator,
    I::Item: Into<&'a Point2>,
```
then we could pass `Vec<Point2>`'s to `points()` without having to move ownership of the underlying data.

A side note: I also was frustrated by the fact that the `view` function signature does not allow it to mutate anything. I was trying to make an `App` where the frame would only be cleared once at the start, and new lines would be drawn on top without erasing old ones. In order to do this I had a list of lines queued to be drawn in my `Model`, and I wanted to draw those lines then clear the list in `view`. I suppose maybe this is bad practice and the idea is to prevent me from doing this. However, for a certain class of drawings where things only accumulate on the screen, this way is optimal and I think we should be enabled to do this.

**Comment by alexfertel** [comment]
@akriegman Maybe I didn't understand something, but `view` is supposed only to display things. A nice analogy is that it is equivalent to the `View` in `MVC`.

There is a separate function that accepts a mutable model called `update`, and it has the signature `fn update(app: &App, model: &mut Model, _update: Update)`.

Wouldn't it be enough to draw the background only in the first frame regarding the clearing issue? Like so:

```rust
    if app.elapsed_frames() == 1 {
        draw.background().color(SKYBLUE);
    }
```

**Comment by alexfertel** [comment]
I think we shouldn't need to move the points too, but I cannot tell if there is some constraint that we are not seeing.

**Comment by mitchmindtree** [comment]
Thanks for the issue!

You can avoid moving the points by doing something like this:

```rust
    draw.polyline()
        .weight(2.)
        .points(model.draw_list.iter().cloned());
```

This feeds the `points` method with an iterator over the `draw_list` that copies one point from the original slice at a time, rather than performing the extra heap allocation as would occur when cloning the whole `Vec`.

In case you're interested, the `Draw` API works by building one big internal mesh whose buffers are re-used between each draw. When `draw.to_frame` is called, a set of commands are encoded that describe how to draw everything (including switching between different pipelines for different blend modes, switching between bind groups for different textures, etc). When the `Frame` is `drop`ped at the end of the `view` function (_or_ when `frame.submit()` is called), only then does the command buffer get submitted to the GPU all at once. This is why you'll rarely see `draw` API methods that take a direct slice of vertices - we try to minimize the number of GPU command buffers submitted, as synchronisation between the CPU and GPU tends to be a much greater performance bottleneck than copying some data on the CPU.

> However, for a certain class of drawings where things only accumulate on the screen, this way is optimal and I think we should be enabled to do this.

You're right about our intentions to generally discourage mutation of the model during `view`, however I think you're also right that occasionally you might run into a situation where you want to re-use some resource or cache something in the `view` itself! Normally in these situations, if I can't achieve my update in the `update` function for whatever reason, I reach for [`RefCell`](https://doc.rust-lang.org/std/cell/struct.RefCell.html) though only around the resources/data that I need to update in the `view` so it's clear what remains immutable and what doesn't. I think there might be some examples of this in the `wgpu_*` examples as this tends to be more common when working with custom pipelines.

Hopefully this helps to clarify nannou's approach a little!

**Comment by akriegman** [comment]
Thanks for the reply. It's good to know there's a reason that the API was made this way, although I don't fully understand it. I'll use that solution, but I just wish it was easier to find :,(
I ended up using `RefCell` for just that purpose. (I found out about it by looking at the Nannou source code, because I realised functions such as `app.set_loop_mode()` shouldn't be able to change the state of `app` through an immutable reference.) The only problem with this solution is `RefCell` comes with some performance overhead (I think, though I'm not sure how much). In any case that extra cost is acceptable for my purposes right now.

**Comment by akriegman** [comment]
> @akriegman Maybe I didn't understand something, but `view` is supposed only to display things. A nice analogy is that it is equivalent to the `View` in `MVC`.
> 
> There is a separate function that accepts a mutable model called `update`, and it has the signature `fn update(app: &App, model: &mut Model, _update: Update)`.

Yeah I get that, the only thing is if I want to create a structure during update, draw it once, then delete it, I need a guarantee that there is one `update()` call for one `view()` call. I tihnk that may be the case, but I couldn't find a guarantee of that anywhere.

**Comment by mitchmindtree** [comment]
> I need a guarantee that there is one update() call for one view() call

I can confirm that is the case. I *think* this is mentioned in the tutorials, not so sure about the docs, but happy to accept a PR that makes this clearer!
